gms/gossiper.cc

```diff
@@ -1708,13 +1712,58 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     utils::chunked_vector<application_state> changed;
     auto&& remote_map = remote_state.get_application_state_map();
 
+    auto revert_to_old_app_state = [&] {
+        try {
+            logger.warn("Started to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+            local_state = old_state;
+            replicate(addr, old_state).get();
+            logger.warn("Finished to roll back to old app state for node={}, old_state={}, status={}, replicate_failed={}, listener_failed={}",
+                    addr, old_state, get_gossip_status(addr), replicate_failed, listener_failed);
+        } catch (...) {
+            logger.warn("Failed to roll back to old app state for node={}, old_state={}, status={}: {}",
+                    addr, old_state, get_gossip_status(addr), std::current_exception());
+            bool critical_app_changed = false;
+            for (auto& app : changed) {
+                if (app == application_state::LOAD ||
+                    app == application_state::CACHE_HITRATES ||
+                    app == application_state::VIEW_BACKLOG) {
+                    continue;
```

Comment:
This "if" is unsafe in case of extension. We should use a switch without a default case so that compilation fails unless every state is explicitly assigned here.

Reply:
Well, the switch trick indeed can help extension.